### PR TITLE
Windows fixes

### LINKED ISSUE #4: Successful build on Windows?
Has this project ever been built on Windows?

### PR DESCRIPTION
**Pull request by caron**
#4 this should address building on windows with visual studio 2013

### CODE REVIEW

File: hdr.cpp [hunk 1]
```diff
@@ -65,10 +65,10 @@ void ExposureSeries::merge() {
 		cout << "Only one exposure was specified -- not doing HDR merging." << endl;
 
 		#pragma omp parallel for
-		for (size_t y=0; y<height; ++y) {
+		for (int y=0; y<height; ++y) {
```
Comment (wjakob) [inline hdr.cpp]: If you change this from size_t to int, then please also add a (size_t) cast at the places where the variable is used.

For instance ``y*width`` can overflow for truly big images (happened to me!),  write ``(size_t) y * width``

File: hdr.cpp [hunk 2]
```diff
@@ -77,15 +77,15 @@ void ExposureSeries::merge() {
 
 	cout << "Merging " << size() << " exposures .." << endl;
 	#pragma omp parallel for
-	for (size_t y=0; y<height; ++y) {
+	for (int y=0; y<height; ++y) {
 		uint32_t offset = y * width;
```
Comment (wjakob) [inline hdr.cpp]: ditto here

Comment (wjakob) [inline hdr.cpp]: You know what -- never mind. Let's leave all of these variables as ``size_t``. This is really an issue with the compiler which does not fully realize OpenMP 3.0, where iteration variables can be unsigned. In that case, I'd rather have a pragma or CMake compiler flag that shuts up any warnings.

**Comment by wjakob** [comment]
Great, thanks! -- I'm adding comments to your PR in a few places.

File: misc.cpp [hunk 3]
```diff
@@ -97,6 +97,7 @@ std::istream& operator>>(std::istream& in, EColorMode& unit) {
 }
 
 int getProcessorCount() {
-	return sysconf(_SC_NPROCESSORS_CONF);
+	//return sysconf(_SC_NPROCESSORS_CONF);
+	return std::thread::hardware_concurrency();
```
Comment (wjakob) [inline misc.cpp]: nice!

File: hdrmerge.h [hunk 4]
```diff
@@ -6,6 +6,9 @@
 #include <map>
 #include <stdexcept>
 #include <cmath>
+#ifdef WIN32
```
Comment (wjakob) [inline hdrmerge.h]: no need for the #ifdef. Let's just always include algorithm.

File: CMakeLists.txt [hunk 5]
```diff
@@ -6,32 +6,36 @@ set(CMAKE_BUILD_TYPE Release)
 #set(CMAKE_CXX_COMPILER /usr/local/bin/g++-4.8)
 
 set(CMAKE_MODULE_PATH ${CMAKE_MODULE_PATH} "${CMAKE_SOURCE_DIR}/misc/")
-find_package(LibXml2 REQUIRED)
```
Comment (wjakob) [inline CMakeLists.txt]: Can you please leave LibXml2? It is needed on Linux when linking with EXIV2 (otherwise you'll get errors about missing symbols)

Comment (wjakob) [inline CMakeLists.txt]: Maybe just wrap it in
```
IF("${CMAKE_SYSTEM}" MATCHES "Linux")
```

File: CMakeLists.txt [hunk 6]
```diff
@@ -6,32 +6,36 @@ set(CMAKE_BUILD_TYPE Release)
...
 find_package(JPEG)
-find_package(Boost 1.40 COMPONENTS program_options REQUIRED )
+find_package(Boost 1.40 COMPONENTS program_options filesystem REQUIRED )
 
 if (OPENMP_FOUND)
 	set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${OpenMP_C_FLAGS}")
 	set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${OpenMP_CXX_FLAGS}")
 endif()
 
-include_directories(${LIBXML2_INCLUDE_DIR})
```
Comment (wjakob) [inline CMakeLists.txt]: this also needs to stay

File: CMakeLists.txt [hunk 7]
```diff
@@ -6,32 +6,36 @@ set(CMAKE_BUILD_TYPE Release)
...
 
 # Compile with C++11 features
 set(CMAKE_CXX_FLAGS "-std=c++11 -Wall ${CMAKE_CXX_FLAGS}")
 
-add_subdirectory(rawspeed)
+add_definitions(-D_USE_MATH_DEFINES -D_UNICODE -DUNICODE)
+
+#add_subdirectory(rawspeed)
 add_executable(hdrmerge input.cpp output.cpp main.cpp hdr.cpp fitexp.cpp resample.cpp misc.cpp)
-target_link_libraries(hdrmerge rawspeed ${LIBXML2_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT}
```
Comment (wjakob) [inline CMakeLists.txt]: ... and this.

File: hdrmerge.h [hunk 8]
```diff
@@ -55,9 +55,9 @@ struct Exposure {
 	std::string toString() const {
 		char buf[10];
 		if (exposure < 1)
-			snprintf(buf, sizeof(buf), "1/%.4g", 1/exposure);
```
Comment (wjakob) [inline hdrmerge.h]: Please leave the snprintf the way it was. You can add

```
#if defined(_WIN32)
#define snprintf _snprintf
#endif
```

File: input.cpp [hunk 9]
```diff
@@ -215,7 +245,10 @@ void ExposureSeries::load() {
 
 	#pragma omp parallel for schedule(dynamic, 1)
 	for (int i=0; i<(int) exposures.size(); ++i) {
-		FileReader f((char *) exposures[i].filename.c_str());
+		//FileReader f((char *)exposures[i].filename.c_str());
+		wchar_t wresult[1024];
```
Comment (wjakob) [inline input.cpp]: Please ``#ifdef _WIN32`` the wide char-related parts here.

File: platform.h [hunk 10]
```diff
@@ -0,0 +1,15 @@
...
+#define __PLATFORM_H
+
+#ifdef _MSC_VER
+#define F_OK 0
+#endif
+
+// platform specific function for snprintf
+#if defined(_MSC_VER) && _MSC_VER < 1900
+
+#define snprintf(buf,len, format,...) _snprintf_s(buf, len,len, format, __VA_ARGS__)
```
Comment (wjakob) [inline platform.h]: Aha, never  mind my earlier comment then :)

**Comment by wjakob** [comment]
Ok, I made it to the end -- some of my earlier comments are invalidated by your last commit. For the other ones, could you please still make the required changes?

**Comment by caron** [comment]
I will address these comments and I have some other small cleanups in mind... like documentation update for building etc.

File: CMakeLists.txt (hunk 5 above)
Comment (caron) [inline CMakeLists.txt]: so I used the exiv2 build server libraries and avoided needing to link libxml2 again..

http://exiv2.dyndns.org:8080/userContent/builds/Categorized/Platform/msvc/

Comment (wjakob) [inline CMakeLists.txt]: Right.. but when you use the distribution-provided libxml e.g. on Linux, you'll get a linker error in this case.

**Comment by caron** [comment]
@wjakob Did these last two commits address your requirements for accepting the PR?

**Pull request merged**